### PR TITLE
Dense2compress

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented dense2csr and dense2csc conversion.
related to #36

### CODE REVIEW

**Comment by gwtnb** [comment]
autopep8

File: cupy/sparse/compressed.py [hunk 1]
```diff
@@ -65,6 +65,16 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
             if len(data) != len(indices):
                 raise ValueError('indices and data should have the same size')
 
+        elif base.isdense(arg1) and arg1.ndim == 2:
+            if self.format == 'csc':
```
Comment (takagi) [inline cupy/sparse/compressed.py]: This branching should be delegated to the subclasses.

    m = self.dense2csx(arg1)
    self.__init__(m, shape=shape, dtype=dtype)
    return

Comment (unnonouno) [inline cupy/sparse/compressed.py]: OK

File: cupy/sparse/compressed.py [hunk 2]
```diff
@@ -65,6 +65,12 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
             if len(data) != len(indices):
                 raise ValueError('indices and data should have the same size')
 
+        elif base.isdense(arg1) and arg1.ndim == 2:
+            data, indices, indptr = self._convert_dense(arg1)
```
Comment (takagi) [inline cupy/sparse/compressed.py]: Another csr_matrix is created to initialize a csr_matrix, a little entangled. An option is to make cusparse.py independent of cupy.sparse.spmatrix suites, but it is followed with some amount of code change. Too much?

File: cupy/sparse/csr.py [hunk 3]
```diff
@@ -60,6 +62,10 @@ def get(self, stream=None):
         return scipy.sparse.csr_matrix(
             (data, indices, indptr), shape=self._shape)
 
+    def _convert_dense(self, x):
+        m = cusparse.dense2csr(x)
```
Comment (takagi) [inline cupy/sparse/csr.py]: Another one is created to initialize a CSR matrix, a little entangled. An option is to make cusparse.py independent of cupy.sparse.spmatrix suites, but it is followed with some amount of code change. Too much?

File: tests/cupy_tests/sparse_tests/test_csr.py [hunk 4]
```diff
@@ -118,6 +118,34 @@ def test_init_copy_other_scipy_sparse(self):
...
+        n = cupy.sparse.csr_matrix(m)
+        self.assertEqual(n.nnz, 0)
+        self.assertEqual(n.shape, (3, 4))
+        cupy.testing.assert_array_equal(n.data, [])
+        cupy.testing.assert_array_equal(n.indices, [])
+        cupy.testing.assert_array_equal(n.indptr, [0, 0, 0, 0])
+
+    @unittest.skipUnless(scipy_available, 'requires scipy')
+    @testing.numpy_cupy_raises(sp_name='sp')
+    def test_init_dense_invalid_ndim(self):
```
Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_csr.py]: `xp` and `sp` arguments

File: tests/cupy_tests/sparse_tests/test_csc.py [hunk 5]
```diff
@@ -115,6 +115,34 @@ def test_init_copy_other_scipy_sparse(self):
...
+        n = cupy.sparse.csc_matrix(m)
+        self.assertEqual(n.nnz, 0)
+        self.assertEqual(n.shape, (3, 4))
+        cupy.testing.assert_array_equal(n.data, [])
+        cupy.testing.assert_array_equal(n.indices, [])
+        cupy.testing.assert_array_equal(n.indptr, [0, 0, 0, 0, 0])
+
+    @unittest.skipUnless(scipy_available, 'requires scipy')
+    @testing.numpy_cupy_raises(sp_name='sp')
+    def test_init_dense_invalid_ndim(self):
```
Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_csc.py]: xp and sp arguments

File: cupy/cusparse.py [hunk 6]
```diff
@@ -496,3 +496,79 @@ def csr2csc(x):
...
+    indptr = cupy.empty(n + 1, 'i')
+    indices = cupy.empty(nnz, 'i')
+
+    _call_cusparse(
+        'dense2csc', x.dtype,
+        handle, m, n, descr.descriptor,
+        x.data.ptr, m, nnz_per_col.data.ptr,
+        data.data.ptr, indices.data.ptr, indptr.data.ptr)
+    # Note that a desciptor is recreated
+    y = cupy.sparse.csc_matrix((data, indices, indptr), shape=x.shape)
```
Comment (takagi) [inline cupy/cusparse.py]: Unnecessary variable `y`.

File: cupy/cusparse.py [hunk 7]
```diff
@@ -496,3 +496,79 @@ def csr2csc(x):
...
+    indptr = cupy.empty(m + 1, 'i')
+    indices = cupy.empty(nnz, 'i')
+
+    _call_cusparse(
+        'dense2csr', x.dtype,
+        handle, m, n, descr.descriptor,
+        x.data.ptr, m, nnz_per_row.data.ptr,
+        data.data.ptr, indptr.data.ptr, indices.data.ptr)
+    # Note that a desciptor is recreated
+    y = cupy.sparse.csr_matrix((data, indices, indptr), shape=x.shape)
```
Comment (takagi) [inline cupy/cusparse.py]: Unnecessary variable `y`.

File: cupy/sparse/csr.py (hunk 3 above)
Comment (unnonouno) [inline cupy/sparse/csr.py]: Hum. In this case, is it better to change all interfaces in cusparse.py?

Comment (takagi) [inline cupy/sparse/csr.py]: Yes, if we take the case. But I often separate layers too strict, the current implementation is acceptable enough for me. We can get forward with that.

Comment (unnonouno) [inline cupy/sparse/csr.py]: I feel the current implementation is enough simple.

Comment (takagi) [inline cupy/sparse/csr.py]: Yes.

**Comment by takagi** [comment]
LGTM!

**Pull request merged**